File: lib-segment/src/main/java/websiteschema/mpsegment/dict/domain/DomainDictFactory.java
package websiteschema.mpsegment.dict.domain;

import org.apache.commons.logging.*;
import websiteschema.mpsegment.conf.MPSegmentConfiguration;

public class DomainDictFactory {

    private final static DomainDictFactory ins = new DomainDictFactory();
    private final static Log l = LogFactory.getLog("segment");
    private volatile DomainDictionary domainDict;
    private final MPSegmentConfiguration config = MPSegmentConfiguration.getInstance();
    private final boolean loadDomainDictionary = config.isLoadDomainDictionary();

    public static DomainDictFactory getInstance() {
        return ins;
    }

    public final void buildDictionary() {
        if (loadDomainDictionary) {
            DomainDictionary dict = new DomainDictionary();
            initializeDictionaryLoaderThenLoad(dict);
            domainDict = dict;
        } else {
            domainDict = new DomainDictionary();
        }
    }

    public DomainDictionary getDomainDictionary() {
        return domainDict;
    }

    public void initializeDictionaryLoaderThenLoad(DomainDictionary dict) {
        try {
            initializeAndLoad(dict);
        } catch (Exception ex) {
            l.error("Exception thrown when load domain dictionary. " + ex.getMessage());
        }
    }

    private void initializeAndLoad(DomainDictionary dict) {
        String classNames[] = config.getDomainDictLoader().split("[,; ]+");
        for (String className : classNames) {
            DomainDictLoader loader = createDictLoader(className);
            if (null != loader) {
                loader.load(dict);
            }
        }
    }

    private DomainDictLoader createDictLoader(String className) {
        try {
            Class clazz = Class.forName(className);
            DomainDictLoader loader = (DomainDictLoader) clazz.newInstance();
            return loader;
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        return null;
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/dict/domain/DomainDictFactory.scala
package websiteschema.mpsegment.dict.domain

import websiteschema.mpsegment.conf.MPSegmentConfiguration

object DomainDictFactory{
  val instance = new DomainDictFactory()

  def apply() = instance
}

class DomainDictFactory {

    private var domainDict: DomainDictionary = null;
    private val config : MPSegmentConfiguration = MPSegmentConfiguration();
    private val loadDomainDictionary : Boolean = config.isLoadDomainDictionary();

    def buildDictionary() {
        if (loadDomainDictionary) {
            val dict = new DomainDictionary();
            initializeDictionaryLoaderThenLoad(dict);
            domainDict = dict;
        } else {
            domainDict = new DomainDictionary();
        }
    }

    def getDomainDictionary() : DomainDictionary = {
        return domainDict;
    }

    def initializeDictionaryLoaderThenLoad(dict: DomainDictionary) {
        try {
            initializeAndLoad(dict);
        } catch {
          case ex: Throwable =>
            System.err.println("Exception thrown when load domain dictionary. " + ex.getMessage());
        }
    }

    private def initializeAndLoad(dict: DomainDictionary) {
        val classNames = config.getDomainDictLoader().split("[,; ]+");
        for (className <- classNames) {
            val loader = createDictLoader(className);
            if (null != loader) {
                loader.load(dict);
            }
        }
    }

    private def createDictLoader(className: String) : DomainDictLoader = {
        try {
            val clazz = Class.forName(className);
            val loader = clazz.newInstance();
            return loader.asInstanceOf[DomainDictLoader];
        } catch {
          case ex: Throwable =>
            ex.printStackTrace()
        }
        return null;
    }
}
